### PR TITLE
Set fixes for tau reco at miniAOD

### PR DESCRIPTION
**Pull request by mbluj**
#### PR description:

This PR contains a set of fixes improving agreement between taus reconstructed with miniAOD inputs and default tau reconstruction with AOD inputs. The fixes are:
* Correct retrieval of track information contained in pat::PackedCandidates in TauChargedHadronProducer;
* Retrieval of individual ECAL and HCAL energies from energy fractions in pat::PackedCandidates in a helper function used by tauIDs and a tau energy algorithm.

Changes are not expected in official production workflows as updated code deals only with, still experimental, tau reconstruction with miniAOD inputs. The tau reconstruction with miniAOD inputs is run as part of tau reco unit tests and is defined by `RecoTauTag/RecoTau/test/rerunTauRecoOnMiniAOD.py`
 

#### PR validation:

Reco tau unit tests passed, in particular one defined in `RecoTauTag/RecoTau/test/rerunTauRecoOnMiniAOD.py`

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoTauTag/RecoTau/plugins/PFRecoTauEnergyAlgorithmPlugin.cc [hunk 1]
```diff
@@ -344,13 +349,29 @@ namespace reco {
...
               if (edm::isFinite(pfCand->ecalEnergy()))
                 allNeutralsSumEn += pfCand->ecalEnergy();
               if (edm::isFinite(pfCand->hcalEnergy()))
                 allNeutralsSumEn += pfCand->hcalEnergy();
               if (edm::isFinite(pfCand->hoEnergy()))
                 allNeutralsSumEn += pfCand->hoEnergy();
+            } else {
+              // TauReco@MiniAOD: individual ECAL and HCAL energies recovered from fractions. Info on HO energy is not (yet?) available in miniAOD.
+              const pat::PackedCandidate* packedCand = dynamic_cast<const pat::PackedCandidate*>(&*signalCand);
+              if (packedCand) {
```
Comment (slava77) [inline RecoTauTag/RecoTau/plugins/PFRecoTauEnergyAlgorithmPlugin.cc:361]: I'd remove this check or add an assert if we get a nullptr here

File: RecoTauTag/RecoTau/plugins/PFRecoTauEnergyAlgorithmPlugin.cc [hunk 2]
```diff
@@ -344,13 +349,29 @@ namespace reco {
...
+                double ecalEn = packedCand->caloFraction() * packedCand->energy() * (1. - packedCand->hcalFraction());
+                double hcalEn = packedCand->caloFraction() * packedCand->energy() * packedCand->hcalFraction();
```
Comment (slava77) [inline RecoTauTag/RecoTau/plugins/PFRecoTauEnergyAlgorithmPlugin.cc:363]:
```suggestion
                auto caloEn = packedCand->caloFraction() * packedCand->energy();
                double hcalEn = caloEn * packedCand->hcalFraction();
                double ecalEn = caloEn - hcalEn;
```
to simplify

File: RecoTauTag/RecoTau/src/PFRecoTauClusterVariables.cc [hunk 3]
```diff
@@ -83,8 +83,16 @@ namespace reco {
...
           ecal_en_in_signal_pf_cands += signal_pfcand->ecalEnergy();
           hcal_en_in_signal_pf_cands += signal_pfcand->hcalEnergy();
+        } else {
+          // TauReco@MiniAOD: individual ECAL and HCAL energies recovered from fractions
+          const pat::PackedCandidate* signal_pcand = dynamic_cast<const pat::PackedCandidate*>(signal_cand.get());
+          if (signal_pcand != nullptr) {
+            ecal_en_in_signal_pf_cands +=
+                signal_pcand->caloFraction() * signal_pcand->energy() * (1. - signal_pcand->hcalFraction());
+            hcal_en_in_signal_pf_cands +=
+                signal_pcand->caloFraction() * signal_pcand->energy() * signal_pcand->hcalFraction();
```
Comment (slava77) [inline RecoTauTag/RecoTau/src/PFRecoTauClusterVariables.cc:93]:
```suggestion
            auto caloEn = signal_pcand->caloFraction() * signal_pcand->energy();
            ecal_en_in_signal_pf_cands += caloEn * (1. - signal_pcand->hcalFraction());
            hcal_en_in_signal_pf_cands += caloEn * signal_pcand->hcalFraction();
```
to reduce repeated calls

File: RecoTauTag/RecoTau/src/PFRecoTauClusterVariables.cc [hunk 4]
```diff
@@ -83,8 +83,16 @@ namespace reco {
         if (signal_pfcand != nullptr) {
           ecal_en_in_signal_pf_cands += signal_pfcand->ecalEnergy();
           hcal_en_in_signal_pf_cands += signal_pfcand->hcalEnergy();
+        } else {
+          // TauReco@MiniAOD: individual ECAL and HCAL energies recovered from fractions
+          const pat::PackedCandidate* signal_pcand = dynamic_cast<const pat::PackedCandidate*>(signal_cand.get());
+          if (signal_pcand != nullptr) {
```
Comment (slava77) [inline RecoTauTag/RecoTau/src/PFRecoTauClusterVariables.cc:89]:
```suggestion
```
remove a check, or add an assert: if we are here and it's not a packedCand, it's most likely that we have a bug

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by mbluj** [comment]
A quick look on logs suggests that failing tests are testing trigger configuration and that the failures are due to problems with accessing input files. Therefore, the failures are not related to this PR.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/32243 	e9bbd0b 
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show no differences (the comparison jobs do not cover the modified code)

**Comment by qliphy** [comment]
+1

**Pull request merged**